File: laravel/contacts/app/Http/Controllers/ContactsController.php
<?php

namespace App\Http\Controllers;

use Illuminate\Http\Request;
use Illuminate\View\View;
//use Illuminate\Routing\Redirector;
use App\Http\Controllers\Controller;
use App\Models\Contact;
use App\Mustache;

use Psr\Log\LoggerInterface;


class ContactsController extends Controller
{
    public function __construct(
        protected Mustache $mustache,
        protected LoggerInterface $logger,
    ) 
    {}

    public function contacts(Request $request): string
    {
        $this->logger->info("CONTACTS");

        $search = $request->query->get('q');

        $token = csrf_token(); //$request->session()->token();

        $contacts;

        if($search)
        {
            $contacts = Contact::where("first", "LIKE", $search."%")
                ->orwhere("last", "LIKE", $search."%")
                ->get();
        }
        else
        {
            $contacts = Contact::all();
        }

        $count = Contact::count();

        $this->logger->info($contacts);

        return  $this->mustache->render( 'index.tpl', [
            "contacts" => $contacts,
            "count" => $count,
            "token" => $token
        ]);

    }

    public function count(Request $request): string
    {
        $this->logger->info("COUNT");

        $count = Contact::count();

        return $count;
    }

    public function show(string $id): string
    {

        $this->logger->info("SHOW");

        $contact = Contact::find($id);

        return  $this->mustache->render( 'show.tpl', $contact );
    }

    public function email(Request $request,string $id): string
    {
        $email = $request->get("email");

        $this->logger->info("EMAIL");

        $existing = Contact::where('email', $email)->first();

        if($existing)
        {
            if($existing->id != $id)
            {
                return "Email already taken!";
            }
        }
        return  "";
    }

    public function create(Request $request): string
    {

        $this->logger->info("CREATE");

        $token = csrf_token(); 

        return  $this->mustache->render( 'new.tpl', [
            "contact" => null,
            "token" => $token
        ] );
    }

    public function insert(Request $request): string
    {
        $email = $request->get("email");
        $first = $request->get("first_name");
        $last = $request->get("last_name");
        $phone = $request->get("phone");

        $this->logger->info("INSERT ".$email);

        $contact = new Contact();
        $contact->email = $email;
        $contact->first = $first;
        $contact->last = $last;
        $contact->phone = $phone;

        $errors = array();
        if(!$email)
        {
            $errors["email"] = "Email is required.";
        }
        if(!$first)
        {
            $errors["first"] = "First name is required.";
        }

        $existing = Contact::where('email', $email)->first();

        if($existing)
        {
            $errors["email"] = "Email must be unique.";
        }

        if(count($errors))
        {
            $token = csrf_token(); 

            $viewdata = array(                
                "errors" => $errors,
                "contact" => $contact,
                "token" => $token
            );
    
            return $this->mustache->render("new.tpl",$viewdata);
        }        

        $contact->save();

        return redirect("/contacts");
    }

    public function edit(Request $request, $id): string
    {
        $this->logger->info("EDIT");

        $contact = Contact::find($id);

        $token = csrf_token(); 

        return  $this->mustache->render( 'edit.tpl', [
            "contact" => $contact,
            "token" => $token
        ] );
    }


    public function update(Request $request, $id): string
    {
        $email = $request->get("email");
        $first = $request->get("first_name");
        $last = $request->get("last_name");
        $phone = $request->get("phone");

        $this->logger->info("UPDATE ".$email);

        $contact = Contact::find($id);

        $contact->email = $email;
        $contact->first = $first;
        $contact->last = $last;
        $contact->phone = $phone;

        $errors = array();
        if(!$email)
        {
            $errors["email"] = "Email is required.";
        }
        if(!$first)
        {
            $errors["first"] = "First name is required.";
        }

        $existing = Contact::where('email', $email)->first();

        if($existing && $existing->id != $id )
        {
            $errors["email"] = "Email must be unique.";
        }

        if(count($errors))
        {
            $token = csrf_token(); 

            $viewdata = array(                
                "errors" => $errors,
                "contact" => $contact,
                "token" => $token
            );
    
            return $this->mustache->render("edit.tpl",$viewdata);
        }        

        $contact->save();

        return redirect("/contacts");
    }

    public function remove(Request $request, string $id)
    {

        $this->logger->info("DELETE: ".$id);

        $contact = Contact::find($id);
        $contact->delete();

        if( $request->headers->get('HX-Trigger') == 'delete-btn')
            return redirect('/contacts', 303);

        return response("")
            ->header('HX-Trigger', "recountEvent");
    }

}

File: spring/contacts/src/main/java/org/oha7/contacts/ContactsController.java
package org.oha7.contacts;

import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.ResponseBody;
import org.springframework.web.servlet.ModelAndView;
import org.springframework.web.servlet.view.RedirectView;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;

import java.util.HashMap;
import java.util.List;

@Controller
public class ContactsController {

    private ContactsRepository repository;

    ContactsController(ContactsRepository repo) {
        this.repository = repo;
    }

    @GetMapping("/contacts")
    public ModelAndView contacts(@RequestParam(required = false) String q) {

        List<Contact> contacts = null;
        
        if(q == null)
        {
            contacts = repository.findAll();
        }
        else
        {
            contacts = repository.search(q + "%");
        }

        var model = new ContactsViewModel (contacts,q,contacts.size());

        return new ModelAndView("index","view",model);
    }

    @GetMapping("/contacts/{id}")
    public ModelAndView show(@PathVariable(value="id") Long id) {

        var contact =repository.findById(id).get();

        var errors = new ContactErrors();
        var model = new ContactViewModel(contact,errors);

        return new ModelAndView("show","view",model);
    }

    @GetMapping("/contacts/new")
    public ModelAndView create() {

        var contact = new Contact();

        var errors = new ContactErrors();
        var model = new ContactViewModel(contact,errors);

        return new ModelAndView("new","view",model);
    }

    @PostMapping("/contacts/new")
    public ModelAndView insert(Contact contact) {

        var errors = validateContact(contact);

        if(errors.hasErrors())
        {
            var model = new ContactViewModel(contact,errors);

            return new ModelAndView("new","view",model);    
        }

        repository.save(contact);

        return new ModelAndView("redirect:/contacts");       
    }

    @GetMapping("/contacts/{id}/edit")
    public ModelAndView edit(@PathVariable(value="id") Long id) {

        var contact =repository.findById(id).get();

        var errors = new ContactErrors();
        var model = new ContactViewModel(contact,errors);

        return new ModelAndView("edit","view",model);
    }

    @PostMapping("/contacts/{id}/edit")
    public ModelAndView update(Contact contact) {

        var errors = validateContact(contact);

        if(errors.hasErrors())
        {
            var model = new ContactViewModel(contact,errors);

            return new ModelAndView("edit","view",model);    
        }

        repository.save(contact);

        return new ModelAndView("redirect:/contacts");       
    }

    @DeleteMapping("/contacts/{id}")
    @ResponseBody
    public Object update( HttpServletResponse response, @PathVariable(value="id") Long id, @RequestHeader(value = "HX-Trigger",required = false) String hxTrigger) {

        repository.deleteById(id);

        if("delete-btn".equals(hxTrigger))
        {
            var view = new RedirectView("/contacts");
            view.setHttp10Compatible(false);
            return view;
        }

        response.setHeader("HX-Trigger", "recountEvent");

        return "";
    }

    @GetMapping("/contacts/{id}/email")
    @ResponseBody
    public String email(@PathVariable(value="id") Long id, @RequestParam(value="email") String email) {

        var contact = repository.findByEmail(email);

        if(contact != null)
        {
            if(contact.id != id)
            {
                return "Email already taken!";
            }
        }
        return "";
    }

    @GetMapping("/contacts/count")
    @ResponseBody
    public String count() {

        var count = repository.count();

        return String.valueOf(count);
    }

    private ContactErrors validateContact(Contact contact) {

        var errors = new ContactErrors();

        if(contact.email.isBlank())
        {
            errors.email = "Email must not be empty.";
        }
        if(contact.last.isBlank())
        {
            errors.last = "Last Name must not be empty.";
        }

        var existing = repository.findByEmail(contact.email);
        if(existing != null && existing.id != contact.id)
        {
            errors.email = "Email must be unique.";
        }
        return errors;
    }
}
